docs/API.md

```diff
@@ -0,0 +1,267 @@
+# API Reference
+
+## Initialization Lifecycle Phases
+
+The following lifecycle phases exist.  Their corresponding event constants are in parentheses.  The source code is in `src/handlers`.
+
+Each lifecycle handler can be provided as an `async` function, or as a Promise, allowing asynchronous execution as necessary.  Note that the application will _wait_ for a phase to be complete before moving on to the next phase.
+
+The corresponding event types are published immediately _after_ the lifecycle phase has completed.  Note that the events are published asynchronously using the [pubsub-js](https://github.com/mroderick/PubSubJS) "publish" method.
+
+### beforeInit (`APP_BEFORE_INIT`)
+
+The `beforeInit` phase has no default behavior.  It can be used to perform actions prior to any of the other phases, but after `App.initialize` has validated its environment configuration.  If you want to perform actions prior to validation of the environment configuration, then write your code before calling `App.initialize` itself.
+
+### configuration (`APP_CONFIGURED`)
+
+The `configuration` phase has no default behavior.
+
+The `configuration` phase can be used to provide dynamic, runtime configuration prior to the initialization of any other services the application may need.
+
+### logging (`APP_LOGGING_CONFIGURED`)
```

Comment:
I wonder if a separate ordered bulleted list of events would make the lifecycle more clear?
I was confused that an event named `APP_LOGGING_CONFIGURED` was coming after an event named `APP_CONFIGURED`.  Maybe you need some clarity in the doc around what `APP_CONFIGURED` means and doesn't mean.

Reply:
Tweaking the name to APP_CONFIG_LOADED to disambiguate from the other ones that say "CONFIGURED" - I struggled with the naming here a bit, admittedly.